hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	typeSuffix               string
+	typeFlag                 astmodel.TypeFlag
+}
+
+func makeEmbeddedResourceRemover(types astmodel.Types) (embeddedResourceRemover, error) {
+	resourceStatusTypes := findAllResourceStatusTypes(types)
+	resourceToSubresourceMap, err := findSubResourcePropertiesTypeNames(types)
+	if err != nil {
+		return embeddedResourceRemover{}, errors.Wrap(err, "couldn't find subresource \"Properties\" type names")
+	}
+
+	resourcePropertiesTypes, err := findAllResourcePropertiesTypes(types)
+	if err != nil {
+		return embeddedResourceRemover{}, errors.Wrap(err, "couldn't find resource \"Properties\" type names")
+	}
+
+	remover := embeddedResourceRemover{
+		types:                    types,
+		resourceToSubresourceMap: resourceToSubresourceMap,
+		resourcePropertiesTypes:  resourcePropertiesTypes,
+		resourceStatusTypes:      resourceStatusTypes,
+		typeSuffix:               "SubResourceEmbedded",
+		typeFlag:                 astmodel.TypeFlag("embeddedSubResource"), // TODO: Instead of flag we could just use a map here if we wanted
+	}
+
+	return remover, nil
+}
+
+func (e embeddedResourceRemover) MakeEmbeddedResourceRemovalTypeVisitor() astmodel.TypeVisitor {
+	visitor := astmodel.MakeTypeVisitor()
+	visitor.VisitObjectType = func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+		typedCtx := ctx.(resourceRemovalVisitorContext)
+
+		if typedCtx.depth <= 2 {
+			// If we are not at sufficient depth, don't bother checking for subresource references. The resource itself and its spec type
+			// will not refer to a subresource. There are some instances of resources (such as Microsoft.Web v20160801 Sites) where the resource
+			// and some child resources reuse the same "Properties" type, which could false some of the logic below
+			return astmodel.IdentityVisitOfObjectType(this, it, ctx)
+		}
+
+		// Before visiting, check if any properties are just referring to one of our sub-resources and remove them
+		subResources := e.resourceToSubresourceMap[typedCtx.resource]
+		for _, prop := range it.Properties() {
+			propTypeName, ok := astmodel.AsTypeName(prop.PropertyType())
+			if !ok {
+				continue
+			}
+
+			// TODO: This is currently no different than the below, but it likely will evolve to be different over time
+			if subResources.Contains(propTypeName) {
+				klog.V(5).Infof("Removing resource %q reference to subresource %q on property %q", typedCtx.resource, propTypeName, prop.PropertyName())
+				it = removeResourceLikeProperties(it)
+				continue
+			}
+
+			if e.resourcePropertiesTypes.Contains(propTypeName) {
+				klog.V(5).Infof("Removing reference to resource %q on property %q", propTypeName, prop.PropertyName())
+				it = removeResourceLikeProperties(it)
+			}
+		}
+
+		return astmodel.IdentityVisitOfObjectType(this, it, ctx)
+	}
+
+	return visitor
+}
+
+func (e embeddedResourceRemover) NewResourceRemovalTypeWalker(visitor astmodel.TypeVisitor) *astmodel.TypeWalker {
```

Comment:
I should write up the situation where you have to snip a different part of the type